python/dgl/dataloading/dataloader.py

```diff
@@ -485,23 +553,103 @@ def create_tensorized_dataset(indices, batch_size, drop_last, use_ddp, ddp_seed)
     def __init__(self, graph, indices, graph_sampler, device='cpu', use_ddp=False,
                  ddp_seed=0, batch_size=1, drop_last=False, shuffle=False,
-                 use_prefetch_thread=False, use_alternate_streams=True, **kwargs):
+                 use_prefetch_thread=None, use_alternate_streams=None,
+                 pin_prefetcher=None, use_uva=False, **kwargs):
+        # (BarclayII) I hoped that pin_prefetcher can be merged into PyTorch's native
+        # pin_memory argument.  But our neighbor samplers and subgraph samplers
+        # return indices, which could be CUDA tensors (e.g. during UVA sampling)
+        # hence cannot be pinned.  PyTorch's native pin memory thread does not ignore
+        # CUDA tensors when pinning and will crash.  To enable pin memory for prefetching
+        # features and disable pin memory for sampler's return value, I had to use
+        # a different argument.  Of course I could change the meaning of pin_memory
+        # to pinning prefetched features and disable pin memory for sampler's returns
+        # no matter what, but I doubt if it's reasonable.
         self.graph = graph
+        self.indices = indices      # For PyTorch-Lightning
+        num_workers = kwargs.get('num_workers', 0)
 
         try:
             if isinstance(indices, Mapping):
                 indices = {k: (torch.tensor(v) if not torch.is_tensor(v) else v)
                            for k, v in indices.items()}
+                indices_device = next(iter(indices.values())).device
             else:
                 indices = torch.tensor(indices) if not torch.is_tensor(indices) else indices
+                indices_device = indices.device
         except:     # pylint: disable=bare-except
             # ignore when it fails to convert to torch Tensors.
             pass
 
+        self.device = _get_device(device)
+
+        # Sanity check - we only check for DGLGraphs.
+        if isinstance(self.graph, DGLHeteroGraph):
+            # Check graph and indices device as well as num_workers
+            if use_uva:
+                if self.graph.device.type != 'cpu':
+                    raise ValueError('Graph must be on CPU if UVM sampling is enabled.')
+                if num_workers > 0:
+                    raise ValueError('num_workers must be 0 if UVM sampling is enabled.')
+
+                # Create all the formats and pin the features - custom GraphStorages
+                # will need to do that themselves.
+                self.graph.create_formats_()
+                self.graph.pin_memory_()
+                for ntype in self.graph.ntypes:
+                    for key, value in self.graph.nodes[ntype].data.items():
+                        self.graph.nodes[ntype].data[key] = value.pin_memory()
+                for etype in self.graph.canonical_etypes:
+                    for key, value in self.graph.edges[etype].data.items():
+                        self.graph.edges[etype].data[key] = value.pin_memory()
```

Comment:
If the crash is not due to in-place `pin_memory_`, should we bring it back? `pin_memory()` method in PyTorch will copy tensors and thus duplicate them in multi-GPU training.

Reply:
I figured out that even if I explicitly call `share_memory_` in the main process before `mp.spawn` I still cannot avoid duplicating the features; the memory consumption and `/dev/shm` consumption still linearly scales with the number of GPUs.  How did you write it so that the memory is not duplicated across processes/GPUs?

My PyTorch version is 1.10.0.

Reply:
Fixed.  For now I noticed that even if the features are not duplicated, the graph itself is not.  I'll leave it for future inspection.  Did you observe the graph itself being duplicated in your own example?  If not, how did you avoid it?

Reply:
Figured it out that the current "proper" way is to first copy the graph into shared memory, then pass the shared memory *name* to the trainer processes, which will get it with `hetero_from_shared_memory`.